File: Kalorie/Kalorie/Core/Utils/FoodItemReporting.swift
//
//  FoodItemReporting.swift
//  Kalorie
//
//  Created by Josef Antoni on 17.09.2026.
//

import Foundation

protocol FoodItemReporting: AnyObject {
    var hasReportedCurrentItem: Bool { get set }
    var isSubmittingReport: Bool { get set }
    var isReportReasonAlertVisible: Bool { get set }
    var reportReasonText: String { get set }
    var alertItem: AlertItem? { get set }
}

extension FoodItemReporting {

    @MainActor
    func loadReportState(barcode: String, fetchMyFoodItemReport: any FetchMyFoodItemReportUseCaseProtocol) async {
        do {
            hasReportedCurrentItem = try await fetchMyFoodItemReport(barcode: barcode) != nil
        } catch {
            Log.warning(error, category: Constants.LogCategory.foodItemReport)
            hasReportedCurrentItem = false
        }
    }

    @MainActor
    func onReportIncorrectDataTapped() {
        guard !hasReportedCurrentItem else { return }
        reportReasonText = ""
        isReportReasonAlertVisible = true
    }

    @MainActor
    func onReportSubmitted(barcode: String, submitFoodItemReport: any SubmitFoodItemReportUseCaseProtocol) async {
        guard !isSubmittingReport else { return }
        isSubmittingReport = true
        defer { isSubmittingReport = false }
        do {
            try await submitFoodItemReport(barcode: barcode, reason: reportReasonText)
            hasReportedCurrentItem = true
        } catch FoodItemReportError.reasonRequired {
            alertItem = AlertItem(title: L10n.FoodItemReport.errorReasonRequired)
        } catch FoodItemReportError.reasonTooLong {
            alertItem = AlertItem(title: L10n.FoodItemReport.errorReasonTooLong)
        } catch {
            Log.error(error, category: Constants.LogCategory.foodItemReport)
            alertItem = AlertItem(title: L10n.Common.errorUnknown)
        }
    }
}

File: Android/app/src/main/java/antoni/kalorie/core/utils/FoodItemReporting.kt
package antoni.kalorie.core.utils

import antoni.kalorie.R
import antoni.kalorie.core.models.FoodItemReportError
import antoni.kalorie.core.usecases.FetchMyFoodItemReportUseCaseProtocol
import antoni.kalorie.core.usecases.SubmitFoodItemReportUseCaseProtocol
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.flow.MutableStateFlow

interface FoodItemReporting {
    val hasReportedCurrentItem: MutableStateFlow<Boolean>
    val isSubmittingReport: MutableStateFlow<Boolean>
    val isReportReasonAlertVisible: MutableStateFlow<Boolean>
    val reportReasonText: MutableStateFlow<String>
    val alertItem: MutableStateFlow<AlertItem?>

    // MARK: - Functions

    suspend fun loadReportState(barcode: String, fetchMyFoodItemReport: FetchMyFoodItemReportUseCaseProtocol) {
        try {
            hasReportedCurrentItem.value = fetchMyFoodItemReport(barcode) != null
        } catch (error: CancellationException) {
            throw error
        } catch (error: Exception) {
            Log.warning(error, Constants.LogCategory.FOOD_ITEM_REPORT)
            hasReportedCurrentItem.value = false
        }
    }

    fun onReportIncorrectDataTapped() {
        if (hasReportedCurrentItem.value) return
        reportReasonText.value = ""
        isReportReasonAlertVisible.value = true
    }

    suspend fun onReportSubmitted(barcode: String, submitFoodItemReport: SubmitFoodItemReportUseCaseProtocol) {
        if (isSubmittingReport.value) return
        isSubmittingReport.value = true
        try {
            submitFoodItemReport(barcode, reportReasonText.value)
            hasReportedCurrentItem.value = true
        } catch (error: CancellationException) {
            throw error
        } catch (_: FoodItemReportError.ReasonRequired) {
            alertItem.value = AlertItem(titleRes = R.string.foodItemReport_error_reasonRequired)
        } catch (_: FoodItemReportError.ReasonTooLong) {
            alertItem.value = AlertItem(titleRes = R.string.foodItemReport_error_reasonTooLong)
        } catch (error: Exception) {
            Log.error(error, Constants.LogCategory.FOOD_ITEM_REPORT)
            alertItem.value = AlertItem(titleRes = R.string.common_error_unknown)
        } finally {
            isSubmittingReport.value = false
        }
    }
}
